wvd-templates/DSC/Functions.ps1

```diff
@@ -327,4 +330,163 @@ function IsRDAgentRegistryValidForRegistration {
+        $sts = $processResult.ExitCode
+
+        $retryCount++
+    } 
+    while ($sts -eq 1618 -and $retryCount -lt 20) # Error code 1618 is ERROR_INSTALL_ALREADY_RUNNING see https://docs.microsoft.com/en-us/windows/win32/msi/-msiexecute-mutex .
+
+    if ($sts -eq 1618) {
+        Write-Log -Message "Stopping retries for $modeAndDisplayName. The last attempt failed with Exit code=$sts"
+        if (-not $isUninstall) {
+            throw "Stopping because $modeAndDisplayName finished with Exit code=$sts"
+        }
+    }
+    else {
+        Write-Log -Message "$modeAndDisplayName finished with Exit code=$sts"
+    }
+} 
+
+<#
+.DESCRIPTION
+Uninstalls any existing RDAgent BootLoader and RD Infra Agent installations and then installs the RDAgent BootLoader and RD Infra Agent using the specified registration token.
+
+.PARAMETER AgentInstallerFolder
+Required path to MSI installer file
+
+.PARAMETER AgentBootServiceInstallerFolder
+Required path to MSI installer file
+
+.PARAMETER StartAgent
+Start the agent service (RDAgentBootLoader) immediately
+#>
+function InstallRDAgents {
+    Param(
+        [Parameter(Mandatory = $true)]
+        [ValidateNotNullOrEmpty()]
+        [string]$AgentInstallerFolder,
+        
+        [Parameter(Mandatory = $true)]
+        [ValidateNotNullOrEmpty()]
+        [string]$AgentBootServiceInstallerFolder,
+    
+        [Parameter(Mandatory = $false)]
+        [ValidateNotNullOrEmpty()]
+        [string]$RegistrationToken,
+    
+        [Parameter(Mandatory = $false)]
+        [ValidateNotNullOrEmpty()]
+        [bool]$StartAgent,
+    
+        [Parameter(mandatory = $false)]
+        [switch]$EnableVerboseMsiLogging
+    )
+
+    # Convert relative paths to absolute paths if needed
+    Write-Log -Message "Boot loader folder is $AgentBootServiceInstallerFolder"
+    $AgentBootServiceInstaller = (Get-ChildItem $AgentBootServiceInstallerFolder\ -Filter *.msi | Select-Object).FullName
+    if ((-not $AgentBootServiceInstaller) -or (-not (Test-Path $AgentBootServiceInstaller))) {
+        throw "RD Infra Agent Installer package is not found '$AgentBootServiceInstaller'"
+    }
+
+    # Convert relative paths to absolute paths if needed
+    Write-Log -Message "Agent folder is $AgentInstallerFolder"
+    $AgentInstaller = (Get-ChildItem $AgentInstallerFolder\ -Filter *.msi | Select-Object).FullName
+    if ((-not $AgentInstaller) -or (-not (Test-Path $AgentInstaller))) {
+        throw "RD Infra Agent Installer package is not found '$AgentInstaller'"
+    }
+
+    if (!$RegistrationToken) {
+        throw "No registration token specified"
+    }
+
+    RunMsiWithRetry -programDisplayName "RDAgentBootLoader" -isUninstall -argumentList @("/x {A38EE409-424D-4A0D-B5B6-5D66F20F62A5}", "/quiet", "/qn", "/norestart", "/passive") -msiOutputLogPath "C:\Users\AgentBootLoaderUnInstall.txt" -msiLogVerboseOutput:$EnableVerboseMsiLogging
+
+    while ($true) {

```

Comment:
What's the point of this while loop ? It looks like it will only run once

Reply:
Per my other comment on [Get-Package](https://docs.microsoft.com/en-us/powershell/module/packagemanagement/get-package?view=powershell-5.1) it will will only return the latest version of a specific installer if there are multiple versions . so this needs to be called multiple times to remove all versions of the installer

Reply:
I did test the case where multiple versions of the RD Agent was installed and this did run multiple iterations and uninstall them all.